lib/libc/sys/setfib.2
.\" Copyright (c) 2008 Cisco Systems.  All rights reserved.
.\"    Author: Julian Elischer
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Neither the name of the Cisco Systems nor the names of its employees
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: releng/9.3/lib/libc/sys/setfib.2 235580 2012-05-18 01:04:03Z gjb $
.\"
.Dd March 19, 2012
.Dt SETFIB 2
.Os
.Sh NAME
.Nm setfib
.Nd set the default FIB (routing table) for the calling process.
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/socket.h
.Ft int
.Fn setfib "int fib"
.Sh DESCRIPTION
The
.Fn setfib
system call sets the associated fib for all sockets opened
subsequent to the call, to be that of the argument
.Fa fib .
The
.Fa fib
argument
must be greater than or equal to 0
and less than the current system maximum which
may be retrieved by the 
.Va net.fibs
sysctl.
The system maximum is set in the kernel configuration file with
.Pp
.Dl options	ROUTETABLES= Ns Em N
.Pp
or in
.Pa /boot/loader.conf
with
.Pp
.Dl net.fibs= Ns Qq Em N
.Pp
where
.Em N
is an integer.
However, this maximum is capped at 16 due to the implementation storing
the fib number in a 4-bit field in
.Xr mbuf 9
flags.
The default fib of the process will be applied to all protocol families 
that support multiple fibs, and ignored by those that do not.
The default fib for a process may be overridden for a socket with the use
of the
.Dv SO_SETFIB
socket option.
.Sh RETURN VALUES
.Rv -std setfib
.Sh ERRORS
The
.Fn setfib
system call
will fail and no action will be taken and return
.Er EINVAL
if the
.Fa fib
argument is greater than the current system maximum.
.Sh SEE ALSO
.Xr setfib 1 ,
.Xr setsockopt 2
.Sh STANDARDS
The
.Fn setfib
system call is a
.Fx
extension however similar extensions
have been added to many other
.Ux
style kernels.
.Sh HISTORY
The
.Fn setfib
function appeared in
.Fx 7.1 .


usr.sbin/setfib/setfib.c
/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 * Copyright (c) 2008 Cisco Systems, All rights reserved
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * 
 * setfib file skelaton taken from nice.c
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/9.3/usr.sbin/setfib/setfib.c 183987 2008-10-17 21:11:09Z delphij $");

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/sysctl.h>

static void usage(void);

int
main(int argc, char *argv[])
{
	long fib = 0;
	int ch;
	char *ep;
	int	numfibs;
	size_t intsize = sizeof(int);

        if (sysctlbyname("net.fibs", &numfibs, &intsize, NULL, 0) == -1)
		errx(1, "Multiple FIBS not supported");
	if (argc < 2)
		usage();
	ep = argv[1];
	/*
	 * convert -N or N to -FN. (N is a number)
	 */
	if (ep[0]== '-' && isdigit((unsigned char)ep[1]))
		ep++;
	if (isdigit((unsigned char)*ep))
               if (asprintf(&argv[1], "-F%s", ep) < 0)
                        err(1, "asprintf");

	while ((ch = getopt(argc, argv, "F:")) != -1) {
		switch (ch) {
		case 'F':
			errno = 0;
			fib = strtol(optarg, &ep, 10);
			if (ep == optarg || *ep != '\0' || errno ||
			    fib < 0 || fib >= numfibs)
				errx(1, "%s: invalid FIB (max %d)",
				    optarg, numfibs - 1);
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 0)
		usage();

	errno = 0;
	if (setfib((int)fib))
		warn("setfib");
	execvp(*argv, argv);
	err(errno == ENOENT ? 127 : 126, "%s", *argv);
}

static void
usage(void)
{

	(void)fprintf(stderr,
	    "usage: setfib [-[F]]value command\n");
	exit(1);
}
